### PR TITLE
Sub models

### PR DESCRIPTION
**Pull request by christophe-david**
This PR implements the idea in #255, that is to provide a way of replacing only a part of a FAST-OAD module.

Usage is described [here](https://fast-oad.readthedocs.io/en/sub-models-complete/documentation/custom_modules/add_submodels.html).

I aimed at having the `RegisterSubModel` class sharing the same base as the existing `RegisterOpenMDAOSystem` and `RegisterPropulsion` classes, so all these classes have been reworked somehow.

Also, this PR has been the occasion to update some Sphinx settings.

### CODE REVIEW

**Review by ScottDelbecq — CHANGES_REQUESTED** [review]
Thanks for this PR! 
I like the way we have chosen. I tested it on parts of the geometry module.
I just have a few comments on the documentation.

File: docs/documentation/custom_modules/add_submodels.rst [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    of the original one.
+
+***************
+Why submodels ?
+***************
+FAST-OAD modules are generally associated to a discipline, and do all the related computations.
+For example, the native weight module computes the masses and the centers of gravity of each
+aircraft part and of the whole aircraft.
+
+Now, let's say we want to modify the computation of wing mass. Then, we could a new weight module
```
[the remark this answers is not in the document]
Comment (christophe-david) [reply docs/documentation/custom_modules/add_submodels.rst:23]: Fixed in 72c49099fe7f957eba0da20eb5c0601aab6eb1cd

File: docs/documentation/custom_modules/add_submodels.rst [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            fuselage_component = CountFuselageAtoms()
+            empennage_component = CountEmpennageAtoms()
+            self.add_subsystem("wing", wing_component, promotes=["*"])
+            self.add_subsystem("fuselage", fuselage_component, promotes=["*"])
+            self.add_subsystem("empennage", empennage_component, promotes=["*"])
+
+
+In the above implementation, someone that would want to provide an alternate method to count
+atoms in the wing, while keeping your method for fuselage, would have to provide its own FAST-OAD
+module, ideally by reusing your :code:`CountFuselageAtoms`class, but possibly by needlessly
```
[the remark this answers is not in the document]
Comment (christophe-david) [reply docs/documentation/custom_modules/add_submodels.rst:73]: Fixed in 72c49099fe7f957eba0da20eb5c0601aab6eb1cd

File: docs/documentation/custom_modules/add_submodels.rst [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+***********************
+It is also possible to deactivate a submodel:
+
+.. code-block:: python
+
+    import fastoad.api as oad
+
+    oad.RegisterSubmodel.active_models["atom_counter.wing"] = None  # The empty string "" is also possible
+
+Then no computation will be done when the "atom_counter.wing" submodel will be called. Of course,
```
[the remark this answers is not in the document]
Comment (christophe-david) [reply docs/documentation/custom_modules/add_submodels.rst:215]: I guess "nothing" will be clearer 😄 
Done in 72c49099fe7f957eba0da20eb5c0601aab6eb1cd

**Review by ScottDelbecq — APPROVED** [review]

**Pull request merged**